gaia/geo/gdal_functions.py

```diff
@@ -176,16 +196,45 @@ def world_to_pixel(geoMatrix, x, y):
         polygon_json = json.dumps(polygon_json)
     poly = ogr.CreateGeometryFromJson(polygon_json)
 
-    # Convert the layer extent to image pixel coordinates
+
     min_x, max_x, min_y, max_y = poly.GetEnvelope()
-    ul_x, ul_y = world_to_pixel(geo_trans, min_x, max_y)
-    lr_x, lr_y = world_to_pixel(geo_trans, max_x, min_y)
+    pixelPoly = world_to_pixel_poly(geo_trans, poly)
+
+    # Convert the layer extent to image pixel coordinates
+    ul_x, lr_x, ul_y, lr_y = pixelPoly.GetEnvelope()
+    ul_x, lr_x, ul_y, lr_y = int(ul_x), int(lr_x), int(ul_y), int(lr_y)
+
+
+    if ul_x < 0:
+        ul_x = 0
+    if ul_y < 0:
+        ul_y = 0
+
+    if lr_x > src_image.RasterXSize:
+        lr_x = src_image.RasterXSize
+
+    if lr_y > src_image.RasterYSize:
+        lr_y = src_image.RasterYSize
 
     # Calculate the pixel size of the new image
+    # Constrain the width and height to the bounds of the image
     px_width = int(lr_x - ul_x)
```

Comment:
probably should be `int(lr_x - ul_x + 1)`, but only if adding the `-1` in computing `lr_x` above.